### PR TITLE
[stdlib] Add a conditional identity cast

### PR DESCRIPTION
**Pull request by Azoy**
This is a conditional variant of `_identityCast` that returns nil if the types do not compare equal.

### CODE REVIEW

**Comment by Azoy** [comment]
@swift-ci please smoke test

File: stdlib/public/core/Builtin.swift [hunk 1]
```diff
@@ -109,6 +109,21 @@ public func _identityCast<T, U>(_ x: T, to expectedType: U.Type) -> U {
   return Builtin.reinterpretCast(x)
 }
 
+/// Returns `x` as its concrete type `U`.
```
Comment (glessard) [inline stdlib/public/core/Builtin.swift:112]:
```suggestion
/// Returns `x` as its concrete type `U`, or `nil` if `x` has a different concrete type.
```

File: stdlib/public/core/Builtin.swift [hunk 2]
```diff
@@ -109,6 +109,21 @@ public func _identityCast<T, U>(_ x: T, to expectedType: U.Type) -> U {
...
+///
+/// This cast can be useful for dispatching to specializations of generic
+/// functions.
+///
+/// - Requires: `x` has type `U`.
```
Comment (glessard) [inline stdlib/public/core/Builtin.swift:117]:
```suggestion
```

**Review by glessard — COMMENTED** [review]
This addition is useful, but I wonder why we can't rely on the dynamic cast's spelling to run as efficiently as this when in the right conditions.

**Comment by lorentey** [comment]
Jinx! I have just made the same addition to Foundation: https://github.com/apple/swift-foundation/pull/92/files#diff-f8876f21e553f16a7614a48f240edc917a50dcf77c3343980462f3ee975b6884R118

We should figure out the best name for this. I think `_conditionalIdentityCast` is too verbose. I went with `_specializingCast` which is meh, but it does rhyme with `@_specialize`, which could help understanding it.

**Comment by lorentey** [comment]
Chatting with @Azoy: perhaps `_specialize(value, for: String.self)` would not be a terrible name for this.

**Comment by Azoy** [comment]
@swift-ci please smoke test

**Review by lorentey — APPROVED** [review]
Looks good to me! It would be good to get a second opinion from someone about the new name, just in case we went too wild. E.g. is there a reason to prefer `_specialize` over `_specialized`? (or `_specializing`??)

(I really like how it reads at the call sites, though.)

**Review by glessard — APPROVED** [review]
`_specialize()` as in the imperative form of the verb is a fine name for a top-level function.

**Comment by dnadoba** [comment]
Just stumbled upon _specialize. I have actually seen it many times while browsing through the Builtin.swift file but never looked to close at the _specialize function as the name didn't make a lot of sense to me. IMHO `_conditionalIdentityCast` is much clearer. This isn't really specializing anything. One use case may be calling out to a specialized function after this function but that is what this function is doing.

**Pull request merged**